contrib/sbtc/token-key-admin/contracts/sbtc-token-key-admin.clar

```diff
@@ -0,0 +1,115 @@
+
+;; title: sbtc-token-key-admin
+;; version:
+;; summary:
+;; description:
+
+;; traits
+;;
+(impl-trait 'SP3FBR2AGK5H9QBDH3EEN6DF8EK8JY7RX8QJ5SVTE.sip-010-trait-ft-standard.sip-010-trait)
+
+;; token definitions
+;; 
+(define-fungible-token sbtc)
+
+;; constants
+;;
+(define-constant contract-owner tx-sender)
+(define-constant err-invalid-caller (err u1))
+(define-constant err-owner-only (err u100))
+(define-constant err-not-token-owner (err u101))
+
+;; data vars
+;;
+(define-data-var coordinator (optional principal) none)
+
+;; data maps
+;;
+(define-map signers uint principal)
+
+;; public functions
+;;
+(define-public (set-coordinator-key (key principal))
+    (if (is-valid-caller)
+        (ok (var-set coordinator (some key)))
+        err-invalid-caller
+    )
+)
+
+(define-public (set-signer-key (id uint) (key principal))
+    (if (is-valid-caller)
+        (ok (map-set signers id key))
+        err-invalid-caller
+    )
+)
+
+(define-public (delete-signer-key (id uint))
+    (if (is-valid-caller)
+        (ok (map-delete signers id))
+        err-invalid-caller
+    )
+)
+
+(define-public (mint! (amount uint))
+    (if (is-valid-caller)
+        (token-credit! tx-sender amount)
+        err-invalid-caller
+    )
+)
+
+(define-public (transfer (amount uint) (sender principal) (recipient principal) (memo (optional (buff 34))))
+	(begin
+		(asserts! (is-eq tx-sender sender) err-not-token-owner)
+		(try! (ft-transfer? sbtc amount sender recipient))
+		(match memo to-print (print to-print) 0x)
+		(ok true)
+	)
+)
+
+;; read only functions
+;;
+(define-read-only (get-coordinator-key)
+    (var-get coordinator)
+)
+
+(define-read-only (get-signer-key (signer uint))
+    (map-get? signers signer)
+)
+
+;;(define-read-only (get-signers)
```

Comment:
I recall you asked about why there was no native support for iteration over maps?  The reason is that Clarity offers the guarantee that the maximum runtime cost of each code path can be computed statically -- i.e. just by inspecting the contract code.  A data map can grow without bound, so the cost to iterate over all of its key/value pairs can also grow unbound.  So, native iteration over data maps is not (currently) supported.

However, there's a straightforward way to do pagination: simply `map` calls to `map-get?` over a list of `signer` values.  For example:

```clarity
(define-constant PAGE_SIZE u5)

(define-private (get-signer-key (signer uint))
   (map-get? signers signer)
)

(define-read-only (get-signer-key-page (page-id uint))
  (let (
     (page-offset (* PAGE_SIZE page-id))
  )
     (map get-signer-key
        (list page-offset 
              (+ u1 page-offset)
              (+ u2 page-offset)
              (+ u3 page-offset)
              (+ u4 page-offset)))
  )
)
```

This way, the VM can determine what the total runtime cost of a call to `get-signer-key-page` will be regardless of how big the map is.

Reply:
As long as the FROST ```key_id```s are monotonically increasing ints from ```1..N```, then yeah it's not hard to iterate over the signers.  @igorsyl wanted to try to remove this requirement, and we totally can on the FROST side, but here it seems problematic.

Reply:
Ah, I wasn't aware.

Can you help me understand the context(s) in which we need to be able to iterate over the list of signers?  If (1) it's for the convenience of downstream clients to learn the list of all signers, and (2) the number of signers is (a) bound and (b) fixed per reward cycle, it might make sense to just store a `(list 4000 ...)` of signers as a data var, so the full list can be queried in one network RPC (along with a Merkle proof).  Growing this list will be comparatively expensive, since the full list must be loaded and stored each time a new entry is added, but it would suffice for now.

Reply:
The context was indeed to prevent multiple read-only calls to the node in order to get the signing keys for every key-id. 

I want to avoid setting the signer keys in bulk because I think people will need to rotate them at irregular intervals.

For now I'll just add a get-signers that iterates over ```[0, num-keys)``` and returns a list.  We won't be calling this frequently enough for that to be a problem.